### ISSUE TITLE
Project name is not validated

### ISSUE THREAD

**Issue by dnicolodi** [pypa/pyproject-metadata#106]
The project name specified in `pyproject.toml` is not validate. Accordingly to https://packaging.python.org/en/latest/specifications/core-metadata/#name it should

**Comment by henryiii** [comment]
I've set this as a 0.9 feature. Don't want to ship this right in the 0.8 rc cycle, and there are questions like where to validate (all other validation is done on the pyproject.toml read), whether to combine the validation errors (would need ExceptionGroups to do properly). I'd be okay with #109 as is though if it gets to sit in `main` for a bit while these things are worked on.

**Comment by dnicolodi** [comment]
I was also unsure about where the best place for validation is. Ideally, an invalid project name would be the first thing reported. However, with the validation as in #109 it gets reported last. I think I added it to the class initialization because it was the easiest thing to do when I implemented it in a subclass downstream in meson-python. Most likely, moving the validation to the `from_pyproject()` constructor is the way to go.
